R/log_shiny_input_changes.R

```diff
@@ -37,35 +38,36 @@ log_shiny_input_changes <- function(
     input,
     namespace = NA_character_,
     excluded_inputs = character(),
-    excluded_pattern = "_width$") {
-  session <- shiny::getDefaultReactiveDomain()
+    excluded_pattern = "_width$",
+    session = shiny::getDefaultReactiveDomain()) {
   if (!(shiny::isRunning() || inherits(session, "MockShinySession") || inherits(session, "session_proxy"))) {
     stop("No Shiny app running, it makes no sense to call this function outside of a Shiny app")
   }
-  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
 
-  # utils::assignInMyNamespace and utils::assignInNamespace are needed
-  # so that observer is executed only once, not twice.
-  input_values <- shiny::isolate(shiny::reactiveValuesToList(input))
-  utils::assignInMyNamespace("shiny_input_values", input_values)
+  if (logger::TRACE > logger::as.loglevel(get_val("TEAL.LOG_LEVEL", "teal.log_level", "INFO"))) {
+    # to avoid setting observers when not needed
+    return(invisible(NULL))
+  }
+
+  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
+  reactive_input_list <- shiny::reactive({
+    input_list <- shiny::reactiveValuesToList(input)
+    input_list[!grepl(excluded_pattern, names(input_list))]
+  })
+  shiny_input_values <- shiny::reactiveVal(shiny::isolate(reactive_input_list()))
 
-  shiny::observe({
-    old_input_values <- shiny_input_values
-    new_input_values <- shiny::reactiveValuesToList(input)
+  shiny::observeEvent(reactive_input_list(), {
+    old_input_values <- shiny_input_values()
+    new_input_values <- reactive_input_list()
     names <- unique(c(names(old_input_values), names(new_input_values)))
-    names <- setdiff(names, excluded_inputs)
-    if (length(excluded_pattern)) {
-      names <- grep(excluded_pattern, names, invert = TRUE, value = TRUE)
-    }
     for (name in names) {
-      old <- old_input_values[name]
-      new <- new_input_values[name]
+      old <- unname(old_input_values[name])
```

Comment:
unname needed because often result of this subset call leads to `old ~ list(<NA> = NULL)` vs `new ~ list(input-name = NULL)`. Even though both are NULL they are not identical.

Reply:
how about `list[[key]]` instead of `unname(list[key])`?

Reply:
it should be a `list`, otherwise `logger` won't be able to print `NULL`

```
logger::log_info("this is { NULL }")
# INFO [2024-09-02 12:50:49] 

logger::log_info("this is { list(NULL) }")
# INFO [2024-09-02 12:50:54] this is NULL
```

Reply:
Thanks! I think I'm able to set this formatter in `.onLoad`

Reply:
Yes. I think that this should be a default formatter when we register the log from any package